### PR TITLE
Add a call scope

### PR DESCRIPTION
**Pull request by yelhamer**
This PR is for adding the call scope.

For the time being, the call scope supports only the features: `API`, `Number`, and `String`.

Also, I have also modified the dynamic-related addresses to include a reference to each one's parent scope/address. This is in order to avoid the mixing of extracted features. For example, two api calls with the same id's, two threads with the same ids, and two processes with the same pid.

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/features/address.py [hunk 1]
```diff
@@ -88,29 +89,54 @@ def __eq__(self, other):
...
+        assert isinstance(other, CallAddress)
+        return (self.thread, self.id) == (other.thread, other.id)
+
+    def __lt__(self, other):
+        assert isinstance(other, CallAddress)
+        return (self.thread, self.id) < (other.thread, other.id)
+
+
+class DynamicReturnAddress(Address):
     """an address from a dynamic analysis trace"""
```
Comment (williballenthin) [inline capa/features/address.py:120]: i dont quite understand what this represents or how its used.

is this used to represent the virtual address of some code in memory during the runtime trace? does it correspond to an unrelocated part of the input sample?

File: capa/features/extractors/base_extractor.py [hunk 2]
```diff
@@ -300,7 +300,7 @@ class ProcessHandle:
     reference to a process extracted by the sandbox.
 
     Attributes:
-        pid: process id
+        address: process' address (pid and pid)
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:303]:
```suggestion
        address: process's address (pid)
```
?

File: capa/features/freeze/__init__.py [hunk 3]
```diff
@@ -194,6 +231,18 @@ class ThreadFeature(HashableModel):
     feature: Feature
 
 
+class CallFeature(HashableModel):
+    """
+    args:
+        call: the call id to which this feature belongs.
+        address: the address at which this feature is found (it's dynamic return address).
```
Comment (williballenthin) [inline capa/features/freeze/__init__.py:238]: > it's dynamic return address

how can a user reference or use this address?

File: capa/rules/__init__.py [hunk 4]
```diff
@@ -188,9 +191,11 @@ def from_dict(self, scopes: dict) -> "Scopes":
     },
     THREAD_SCOPE: {
         capa.features.common.MatchedRule,
-        capa.features.common.String,
         capa.features.common.Substring,
         capa.features.common.Regex,
```
Comment (williballenthin) [inline capa/rules/__init__.py:195]: all of these feature types should be moved to call scope

File: capa/rules/__init__.py [hunk 5]
```diff
@@ -559,6 +569,17 @@ def build_statements(d, scopes: Scopes):
             THREAD_SCOPE, build_statements(d[key][0], Scopes(dynamic=THREAD_SCOPE)), description=description
         )
 
+    elif key == "call":
+        if all(s not in scopes for s in (FILE_SCOPE, PROCESS_SCOPE, THREAD_SCOPE)):
+            raise InvalidRule("thread subscope supported only for the process scope")
```
Comment (williballenthin) [inline capa/rules/__init__.py:574]:
```suggestion
            raise InvalidRule("call subscope supported only for the process and thread scopes")
```

**Review by williballenthin — CHANGES_REQUESTED** [review]
on the whole - very good! its so cool to read a 500 line PR and find yourself nodding along the entire way.

there are a couple places to make small changes that i've referenced inline. once these are addressed, the PR can probably be merged.

i have some personal thoughts on the terminology, but im not sure they are compelling nor important, and they don't block the logic you proposed, so i'll raise this 1:1.

**Comment by williballenthin** [comment]
also, please add some tests that demonstrate the new rule syntax.

File: capa/features/address.py (hunk 1 above)
Comment (williballenthin) [reply capa/features/address.py:120]: i may address this and the call/address comment below in a PR that i propose to you after we merge this PR. i'd like to demonstrate how i think about this and we can compare/contrast.

no need to resolve this comment now.

File: capa/features/freeze/__init__.py (hunk 3 above)
Comment (williballenthin) [reply capa/features/freeze/__init__.py:238]: see above: no need to resolve this comment now.

Comment (yelhamer) [reply capa/features/freeze/__init__.py:238]: > how can a user reference or use this address?

Currently, my interpretation of this is that it's the address associated with that call feature; so for some sandbox, if an argument to an api call is a memory address of a buffer, then the value can be the string in memory, and the address is that in-memory string's location.

Regardless, for CAPE `call` and `address` should be similar, which makes me think that maybe it should be removed for now.

File: capa/features/address.py [hunk 6]
```diff
@@ -88,29 +89,54 @@ def __eq__(self, other):
         return (self.process, self.tid) == (other.process, other.tid)
 
     def __lt__(self, other):
+        assert isinstance(other, ThreadAddress)
         return (self.process, self.tid) < (other.process, other.tid)
 
 
-class DynamicAddress(Address):
+class CallAddress(Address):
```
Comment (williballenthin) [inline capa/features/address.py:96]: i wonder if we should call this `DynamicCallAddress` like we use `DynamicReturnAddress` below?

File: capa/features/extractors/cape/call.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    calls: List[Dict[str, Any]] = process["calls"]
+    call = calls[ch.address.id]
+    assert call["thread_id"] == str(th.address.tid)
+    caller = DynamicReturnAddress(call=ch.address, return_address=int(call["caller"], 16))
+    # list similar to disassembly: arguments right-to-left, call
+    for arg in call["arguments"][::-1]:
+        try:
+            yield Number(int(arg["value"], 16), description=f"{arg['name']}"), caller
+        except ValueError:
+            yield String(arg["value"], description=f"{arg['name']}"), caller
```
Comment (williballenthin) [inline capa/features/extractors/cape/call.py:53]:
```suggestion
            yield Number(int(arg["value"], 16)), caller
        except ValueError:
            yield String(arg["value"]), caller
```

clever use of the `description` kwarg, but today its only meant for rule authors. we could think about extending its interpretation like you do here, but i think that should be a separate discussion.

File: capa/features/address.py (hunk 6 above)
Comment (yelhamer) [reply capa/features/address.py:96]: Hmm, I agree. I had the general future case in mind where the call scope would be also part of static analysis. But given that this address has the thread attribute right now, maybe we should use `DynamicCallAddress`. Perhaps we can introduce a `StaticCallAddress` when the call scope is added to static analysis.

File: capa/main.py [hunk 8]
```diff
@@ -366,24 +367,24 @@ def pbar(s, *args, **kwargs):
     return matches, meta
 
 
-def find_thread_capabilities(
```
Comment (williballenthin) [inline capa/main.py:369]: i wonder if we've reached the point where we should separate out this matching logic into its own namespace, and then we can have subnamespaces for static and dynamic flavors. there's hundreds of lines of logic here, which seems excessive for the main script. we can address this in another PR.

Comment (williballenthin) [reply capa/main.py:369]: #1684

File: tests/test_rules.py [hunk 9]
```diff
@@ -353,6 +353,30 @@ def test_multi_scope_rules_features():
...
+                  - and:
+                    - or:
+                      - api: socket
+                      - and:
+                        - os: linux
+                        - mnemonic: syscall
+                        - number: 41 = socket()
+                    - number: 6 = IPPROTO_TCP
+                    - number: 1 = SOCK_STREAM
+                    - number: 2 = AF_INET
```
Comment (williballenthin) [inline tests/test_rules.py:375]: two spaces per indent here, please. for consistency.

Comment (williballenthin) [reply tests/test_rules.py:375]:
```suggestion
            rule:
              meta:
                name: test rule
                scopes:
                  static: instruction
                  dynamic: call
              features:
                - and:
                  - or:
                    - api: socket
                    - and:
                      - os: linux
                      - mnemonic: syscall
                      - number: 41 = socket()
                  - number: 6 = IPPROTO_TCP
                  - number: 1 = SOCK_STREAM
                  - number: 2 = AF_INET
```

File: tests/test_rules.py [hunk 10]
```diff
@@ -489,12 +513,30 @@ def test_subscope_rules():
...
+                        meta:
+                            name: test call subscope
+                            scopes:
+                                static: basic block
+                                dynamic: thread
+                        features:
+                            - and:
+                                 - string: "explorer.exe"
+                                 - call:
+                                    - api: HttpOpenRequestW
```
Comment (williballenthin) [inline tests/test_rules.py:529]:
```suggestion
                    rule:
                      meta:
                        name: test call subscope
                        scopes:
                          static: basic block
                          dynamic: thread
                      features:
                        - and:
                           - string: "explorer.exe"
                           - call:
                              - api: HttpOpenRequestW
```

File: tests/unsupported_capa_rules.txt
Comment (williballenthin) [inline tests/unsupported_capa_rules.txt:1]: please remove

File: tests/unsupported_capa_rules.yml
Comment (williballenthin) [inline tests/unsupported_capa_rules.yml:1]: please remove

**Review by williballenthin — CHANGES_REQUESTED** [review]
four trivial changes (two formatting, two empty files) and then rename the DynamicCallAddress class. then ready to merge!

**Comment by williballenthin** [comment]
while you're doing that, i'm going to sync master into the dynamic branch. i dont *think* there will be any conflicts, but we'll see shortly.

**Review by williballenthin — APPROVED** [review]
lgtm, thanks @yelhamer

**Pull request merged**